modules/grpc/CMakeLists.txt

```diff
@@ -1,20 +1,42 @@
-set(GRPC_DEPS_FOUND ${ENABLE_CPP})
 
-if (GRPC_DEPS_FOUND)
-  find_package(Protobuf 3.6.1)
+if (NOT ENABLE_CPP AND ENABLE_GRPC)
+    message(FATAL_ERROR "C++ support is mandatory when the GRPC modules are enabled.")
+endif()
+
+set(ENABLE_GRPC_DEFAULT FALSE)
+
+if (NOT APPLE) # TODO: Fix GRPC support on macOS and remove this later
+  set(ENABLE_GRPC_DEFAULT ${ENABLE_CPP})
+endif()
+
+if (ENABLE_GRPC_DEFAULT OR ENABLE_GRPC)
+  set(GRPC_DEPS_FOUND TRUE)
+
+  find_package(Protobuf 3.6.1 QUIET)
   if (NOT Protobuf_FOUND)
+    if (ENABLE_GRPC)
+      message(FATAL_ERROR "ProtoBuf libraries not found.")
+    endif()
     set(GRPC_DEPS_FOUND FALSE)
   endif()
-endif()
 
-if (GRPC_DEPS_FOUND)
-  find_package(gRPC 1.16.1)
-  if (NOT gRPC_FOUND)
-    set(GRPC_DEPS_FOUND FALSE)
+  if (GRPC_DEPS_FOUND)
+    # TODO: Unfortunately find_package does not respect QUIT param here sometimes if the package cannot be found (e.g. on macOS)
+    #       Try to add a proper, silent discovery solution
+    find_package(gRPC 1.16.1 QUIET)
+    if (NOT gRPC_FOUND)
+      if (ENABLE_GRPC)
+        message(FATAL_ERROR "GRPC++ libraries not found.")
+      endif()
+      set(GRPC_DEPS_FOUND FALSE)
+    endif()
   endif()
+
+  set(ENABLE_GRPC_DEFAULT ${GRPC_DEPS_FOUND})
+
 endif()
 
-module_switch(ENABLE_GRPC "Enable GRPC" GRPC_DEPS_FOUND)
+module_switch(ENABLE_GRPC "Enable GRPC" ENABLE_GRPC_DEFAULT)
```

Comment:
What do you think about something like this?
```
if (DEFINED ENABLE_GRPC AND NOT ENABLE_GRPC)
  return()
endif()

set(GRPC_DEPS_FOUND TRUE)

if (NOT ENABLE_CPP)
  if (ENABLE_GRPC)
    message(FATAL_ERROR "C++ support is mandatory when the GRPC modules are enabled.")
  endif()
  set(GRPC_DEPS_FOUND FALSE)
endif()

if (GRPC_DEPS_FOUND)
  find_package(Protobuf 3.6.1 QUIET)
  if (NOT Protobuf_FOUND)
    if (ENABLE_GRPC)
      message(FATAL_ERROR "ProtoBuf libraries not found.")
    endif()
    set(GRPC_DEPS_FOUND FALSE)
  endif()
endif()

if (GRPC_DEPS_FOUND)
  # TODO: Unfortunately find_package does not respect QUIET param here sometimes if the package cannot be found (e.g. on macOS)
  #       Try to add a proper, silent discovery solution
  find_package(gRPC 1.16.1 QUIET)
  if (NOT gRPC_FOUND)
    if (ENABLE_GRPC)
      message(FATAL_ERROR "GRPC++ libraries not found.")
    endif()
    set(GRPC_DEPS_FOUND FALSE)
  endif()
endif()

module_switch(ENABLE_GRPC "Enable GRPC" GRPC_DEPS_FOUND)
```

It makes sure that if someone explicitly disabled `ENABLE_GRPC`, the module is skipped. Otherwise it checks for `ENABLE_CPP`, if it is disabled and `ENABLE_GRPC` is explicitly enabled it throws an error, if it is disabled and `ENABLE_GRPC` is not explicitly enabled it skips the module. The rest is the same: checks a dependency, if found continue, if not found and `ENABLE_GRPC` is explicitly set then throw an error or if not found and `ENABLE_GRPC` is not explicitly set, then skip the module.

The default value of `ENABLE_GRPC` would be nothing in both Apple or other OSs. `ENABLE_CPP` is disabled on Apple by default, but if the user enables it, `ENABLE_GRPC` is still not explicitly enabled, so when the dependencies are not found, grpc will be skipped. If the dependencies are found it tries to build it on Apple, too, but it will fail because of the linker error. I think this is the expected behavior, the user has the capability to explicitly disable GRPC in this case.

I would prefer something like this because the implementation is flat, which is easier to understand IMO, and there are no Apple related conditions, which would further complicate the implementation. Also I managed to skip using an additional variable (`ENABLE_GRPC_DEFAULT`), which makes it less complex.

Reply:
Ohh, after a bunch of rewrites, I successfully skipped the case when `ENABLE_GRPC` is explicitly disabled, excellent :S

As we say "izlesek es pofonok" ;) Even though ENABLE_GRPC_DEFAULT is a plus variable, that makes the code more readable for me as every condition has its own state storage, using `GRPC_DEPS_FOUND` for multiple purposes is a bit confusing for me, but as you wish.

I'd like to add a tiny modification to yours to have the `module_switch` always running as without it `ENABLE_GRPC` will not be enumerated in the summary list
sg. like this

```

set(GRPC_DEPS_FOUND FALSE)

if (NOT DEFINED ENABLE_GRPC OR ENABLE_GRPC)

	set(GRPC_DEPS_FOUND TRUE)

	if (NOT ENABLE_CPP)
	  if (ENABLE_GRPC)
		message(FATAL_ERROR "C++ support is mandatory when the GRPC modules are enabled.")
	  endif()
	  set(GRPC_DEPS_FOUND FALSE)
	endif()

	if (GRPC_DEPS_FOUND)
	  find_package(Protobuf 3.6.1 QUIET)
	  if (NOT Protobuf_FOUND)
		if (ENABLE_GRPC)
		  message(FATAL_ERROR "ProtoBuf libraries not found.")
		endif()
		set(GRPC_DEPS_FOUND FALSE)
	  endif()
	endif()

	if (GRPC_DEPS_FOUND)
	  # TODO: Unfortunately find_package does not respect QUIET param here sometimes if the package cannot be found (e.g. on macOS)
	  #       Try to add a proper, silent discovery solution
	  find_package(gRPC 1.16.1 QUIET)
	  if (NOT gRPC_FOUND)
		if (ENABLE_GRPC)
		  message(FATAL_ERROR "GRPC++ libraries not found.")
		endif()
		set(GRPC_DEPS_FOUND FALSE)
	  endif()
	endif()

endif()

module_switch(ENABLE_GRPC "Enable GRPC" GRPC_DEPS_FOUND)


```

Reply:
Awesome! You can keep the additional variable if you'd like to, having it or not is just preference after all.